### PR TITLE
Resolve multiple same UID definitions with different versions

### PR DESCRIPTION
**Pull request by live1206**
#3607 

When there are multiple same UIDs with different versions, take the latest one respecting the document version

### CODE REVIEW

File: src/docfx/build/xref/XrefMap.cs [hunk 1]
```diff
@@ -300,8 +301,19 @@ private IEnumerable<XrefSpec> ExpandInternalXrefSpecs()
             return loadedInternalSpecs;
         }
 
-        private (InternalXrefSpec spec, Document referencedFile) GetLatestInternalXrefMap(List<(InternalXrefSpec spec, Document referencedFile)> specs)
-            => specs.OrderByDescending(item => item.spec.Monikers.FirstOrDefault(), _context.MonikerProvider.Comparer).FirstOrDefault();
+        private (InternalXrefSpec spec, Document referencedFile) GetLatestInternalXrefMap(List<(InternalXrefSpec spec, Document referencedFile)> specs, List<string> monikers = null)
+        {
+            var orderedSpecs = specs.OrderByDescending(x => x.spec.Monikers.FirstOrDefault(), _context.MonikerProvider.Comparer);
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: can these be pre-ordered?

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: Would there be any difference with pre-ordered? We just sort them once, and the list is fairly small

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: you are right, each resolve would trigger one sort

File: src/docfx/build/xref/XrefMap.cs [hunk 2]
```diff
@@ -300,8 +301,24 @@ private IEnumerable<XrefSpec> ExpandInternalXrefSpecs()
...
+            else
+            {
+                // Firstly check if any spec has monikers intersecting with monikers of the root file
+                // If not, take the spec with highest version whose monikers are all lower than the lowest moniker of the root file
+                var (spec, referencedFile) = specs.FirstOrDefault(x => x.spec.Monikers.Intersect(monikers).Any());
+                if (spec is null)
+                {
+                    (spec, referencedFile) = specs.FirstOrDefault(x => x.spec.Monikers.All(moniker => _context.MonikerProvider.Comparer.Compare(moniker, monikers.FirstOrDefault()) < 0));
+                }
+                return (spec, referencedFile);
```
Comment (928PJY) [inline src/docfx/build/xref/XrefMap.cs]: add a todo for monikers contains multiple  product

File: docs/specs/moniker.yml [hunk 3]
```diff
@@ -453,66 +453,220 @@ outputs:
...
-    No version
-  docs/b.md: |
-    Link to @a
+    Moniker: 
+      - netcore-1.2
+      - netcore-2.0
+  docs/v2/b.md: |
+    Link to @a with netcore-1.2, netcore-2.0
+  docs/v3/b.md: |
+    Link to @a with netcore-1.2, netcore-2.0
```
Comment (928PJY) [inline docs/specs/moniker.yml]: >netcore-1.2, netcore-2.0 [](start = 20, length = 24)

Link to @a with netcore-2.1, netcore-2.2

File: docs/specs/moniker.yml [hunk 4]
```diff
@@ -453,66 +453,220 @@ outputs:
     {"references":[{"uid":"a","href":"https://docs.com/docs/a","name":"netcore-2.0"}]}
 ---
 # Define same uid with multiple moniker ranges, 
-# reference to it without moniker in the same docset, shoule take the latest one
+# reference to it without moniker in the same docset
+# shoule take the latest one but not later than the current document
```
Comment (928PJY) [inline docs/specs/moniker.yml]: >shoule take the latest one but not later than the current document [](start = 2, length = 66)

no later than the max or min of the current document?

File: src/docfx/build/xref/XrefMap.cs [hunk 5]
```diff
@@ -59,7 +59,8 @@ public void OutputXrefMap(Context context)
             string resolvedHref;
             string displayPropertyValue;
             string name;
-            if (TryResolve(uid, href, moniker, out var spec))
+            var (_, rootFileMonikers) = _context.MonikerProvider.GetFileLevelMonikersWithInfo(rootFile);
```
Comment (928PJY) [inline src/docfx/build/xref/XrefMap.cs]: >GetFileLevelMonikersWithInfo [](start = 65, length = 28)

this doesn't contains the monikerRange in the file metadata?

**Comment by live1206** [comment]
covered in https://github.com/dotnet/docfx/pull/4824

**Pull request closed without merging**